src/build_cfg.c

```diff
@@ -0,0 +1,357 @@
+            struct CfVariable *rhs = build_cfg_for_expression(st, &expr->data.operands[1]);
+            assert(same_type(
+                &expr->data.operands[0].type_after_implicit_cast,
+                &expr->data.operands[1].type_after_implicit_cast));
+            bool is_signed = expr->data.operands[0].type_after_implicit_cast.kind == TYPE_SIGNED_INTEGER;
+
+            struct CfInstruction ins;
+            const char *debugname;
+            bool negate = false;
+            bool swap = false;
+
+            switch(expr->kind) {
+                case AST_EXPR_ADD: debugname = "$add"; ins.kind = CF_INT_ADD; break;
+                case AST_EXPR_SUB: debugname = "$sub"; ins.kind = CF_INT_SUB; break;
+                case AST_EXPR_MUL: debugname = "$mul"; ins.kind = CF_INT_MUL; break;
+                case AST_EXPR_DIV: debugname = (is_signed ? "$sdiv" : "$udiv"); ins.kind = (is_signed ? CF_INT_SDIV : CF_INT_UDIV); break;
+                case AST_EXPR_EQ: debugname = "$eq"; ins.kind = CF_INT_EQ; break;
+                case AST_EXPR_NE: debugname = "$ne"; ins.kind = CF_INT_EQ; negate=true; break;
+                case AST_EXPR_LT: debugname = "$lt"; ins.kind = CF_INT_LT; break;
+                case AST_EXPR_GT: debugname = "$gt"; ins.kind = CF_INT_LT; swap=true; break;
+                case AST_EXPR_LE: debugname = "$le"; ins.kind = CF_INT_LT; negate=true; swap=true; break;
+                case AST_EXPR_GE: debugname = "$ge"; ins.kind = CF_INT_LT; negate=true; break;
+                default: assert(0);
+            }
+
+            ins.data.operands[0] = swap?rhs:lhs;
+            ins.data.operands[1] = swap?lhs:rhs;
+
+            ins.destvar = result = add_variable(st, &expr->type_before_implicit_cast, debugname);
+            Append(&st->current_block->instructions, ins);
+
+            if (negate) {
+                struct CfVariable *result2 = add_variable(st, &boolType, debugname);
+                Append(&st->current_block->instructions, (struct CfInstruction){
+                    .kind = CF_BOOL_NEGATE,
+                    .data.operands = {result},
+                    .destvar = result2,
+                });
+                result = result2;
+            }
+        }
+    }
+
+    if (result == NULL) {
+        // call to function with '-> void'
+        assert(expr->kind == AST_EXPR_CALL);
+        return NULL;
+    }
+
+    assert(same_type(&result->type, &expr->type_before_implicit_cast));
+    return build_cfg_for_implicit_cast(st, result, &expr->type_after_implicit_cast);
+}
+
+static struct CfVariable *build_cfg_for_address_of_expression(const struct State *st, const struct AstExpression *address_of_what)
+{
+    switch(address_of_what->kind) {
+    case AST_EXPR_GET_VARIABLE:
+        {
+            struct CfVariable *var = find_variable(st, address_of_what->data.varname);
+            // TODO: shouldn't need to create a new type here
```

Comment:
Or maybe it would be better to clone all types instead of referencing to them in AST?